### PR TITLE
PIN-4916: importEservice endpoint

### PR DESCRIPTION
**Pull request by AsterITA**
https://pagopa.atlassian.net/browse/PIN-4916

### CODE REVIEW

File: src/main/resources/interface-specification.yml [hunk 1]
```diff
@@ -3927,6 +3927,78 @@ paths:
...
+              description: Max allowed requests within time interval
+            'X-Rate-Limit-Remaining':
+              schema:
+                type: integer
+              description: Remaining requests within time interval
+            'X-Rate-Limit-Interval':
+              schema:
+                type: integer
+              description: Time interval in milliseconds. Allowed requests will be constantly replenished during the interval. At the end of the interval the max allowed requests will be available
+        '409':
```
Comment (galales) [inline src/main/resources/interface-specification.yml:3983]: in which case do we return conflict state?

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala [hunk 2]
```diff
@@ -1069,4 +1029,184 @@ final case class EServicesApiServiceImpl(
...
+
+    val result: Future[CreatedResource] = for {
+      tenantId         <- getOrganizationIdFuture(contexts)
+      zipFile          <- fileManager.getFile(ApplicationConfiguration.importEServiceContainer)(
+        s"${ApplicationConfiguration.importEServicePath}/$tenantId/${fileResource.filename}"
+      )
+      zipPath          <- extractZipToTempDirectory(zipFile, tenantId).toFuture
+      importedEservice <- checkZipStructure(zipPath).toFuture.recoverWith { case ex: Throwable =>
+        deleteTempDirectory(zipPath)
+        throw ex
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala]: `recoverWith` expects a `Future` in return, you can use it to raise the error again.
(throwing error is avoided when using effects)
```suggestion
        Future.failed(ex)
```

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala [hunk 3]
```diff
@@ -1069,4 +1029,184 @@ final case class EServicesApiServiceImpl(
...
+      }
+      eserviceSeed = CatalogProcess.EServiceSeed(
+        name = importedEservice.name,
+        description = importedEservice.description,
+        technology = importedEservice.technology.toProcess,
+        mode = importedEservice.mode.toProcess
+      )
+      eService <- catalogProcessService.createEService(eserviceSeed).recoverWith { case ex: Throwable =>
+        deleteTempDirectory(zipPath)
+        throw ex
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala]: same
```suggestion
        Future.failed(ex)
```

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala [hunk 4]
```diff
@@ -1069,4 +1029,184 @@ final case class EServicesApiServiceImpl(
...
+        dailyCallsTotal = importedEservice.descriptor.dailyCallsTotal,
+        agreementApprovalPolicy = importedEservice.descriptor.agreementApprovalPolicy.toProcess,
+        attributes = CatalogProcess.AttributesSeed(certified = Seq.empty, declared = Seq.empty, verified = Seq.empty)
+      )
+      descriptor <- catalogProcessService.createDescriptor(eService.id, descriptorSeed).recoverWith {
+        case ex: Throwable =>
+          deleteTempDirectory(zipPath)
+          for {
+            _ <- catalogProcessService.deleteEService(eService.id)
+          } yield throw ex
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala]: minor, we can avoid for comprehension here
```suggestion
          catalogProcessService.deleteEService(eService.id).flatMap(_ => Future.failed(ex))
```

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala [hunk 5]
```diff
@@ -1069,4 +1029,184 @@ final case class EServicesApiServiceImpl(
...
+      } yield ()
+
+      result.recoverWith { case ex: Throwable =>
+        deleteTempDirectory(zipPath)
+        for {
+          _ <- descriptor.docs.traverse { doc =>
+            catalogProcessService.deleteEServiceDocumentById(eService.id, descriptor.id, doc.id)
+          }
+          _ <- catalogProcessService.deleteEService(eService.id)
+          _ <- catalogProcessService.deleteDraft(eService.id, descriptor.id)
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala]: this is already done by `deleteEService`
```suggestion
```

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala [hunk 6]
```diff
@@ -1069,4 +1029,184 @@ final case class EServicesApiServiceImpl(
...
+          fileType,
+          descriptor.id,
+          uuidSupplier
+        )
+      } yield ()
+
+      result.recoverWith { case ex: Throwable =>
+        deleteTempDirectory(zipPath)
+        for {
+          _ <- descriptor.docs.traverse { doc =>
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala:1149]: we need to delete also the interface

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala [hunk 7]
```diff
@@ -1069,4 +1029,184 @@ final case class EServicesApiServiceImpl(
...
+      descriptor: CatalogProcess.EServiceDescriptor,
+      zipPath: Path,
+      file: ImportedDoc,
+      fileType: String
+    ): Future[Unit] = {
+      val filePath = zipPath.resolve(file.path).toString
+      val fileName = Paths.get(file.path).getFileName.toString
+
+      val result = for {
+        mimeType    <- Try(new Tika().detect(readFileAsBytes(filePath), fileName)).toEither.toFuture
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala]:
```suggestion
        mimeType    <- Try(new Tika().detect(readFileAsBytes(filePath), fileName)).toFuture
```

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala [hunk 8]
```diff
@@ -1069,4 +1029,184 @@ final case class EServicesApiServiceImpl(
...
+
+      for {
+        jsonContent      <- Either
+          .catchNonFatal(Files.readString(path.resolve("configuration.json")))
+          .leftMap(ex => InvalidZipStructure(directoryName, "Error reading configuration.json: " + ex.getMessage))
+        importedEservice <- Try(jsonContent.parseJson.convertTo[ImportedEservice]) match {
+          case Success(eservice) => Right(eservice)
+          case Failure(ex)       =>
+            Left(InvalidZipStructure(directoryName, "Error decoding configuration.json: " + ex.getMessage))
+        }
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala]: very minor, just an alternative
```suggestion
        importedEservice <- Try(jsonContent.parseJson.convertTo[ImportedEservice]).toEither.leftMap(ex =>
          InvalidZipStructure(directoryName, "Error decoding configuration.json: " + ex.getMessage)
        )
```

**Review by galales — COMMENTED** [review]
🚀 few code sugar suggestions and a couple of points to address

File: src/main/resources/interface-specification.yml (hunk 1 above)
Comment (AsterITA) [reply src/main/resources/interface-specification.yml:3983]: Added the `recoverWith` needed for the bubble up from the process for this case

File: src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala [hunk 9]
```diff
@@ -1151,8 +1149,10 @@ final case class EServicesApiServiceImpl(
           _ <- descriptor.docs.traverse { doc =>
             catalogProcessService.deleteEServiceDocumentById(eService.id, descriptor.id, doc.id)
           }
+          _ <- descriptor.interface.traverse(interface =>
+            catalogProcessService.deleteEServiceDocumentById(eService.id, descriptor.id, interface.id)
+          )
           _ <- catalogProcessService.deleteEService(eService.id)
-          _ <- catalogProcessService.deleteDraft(eService.id, descriptor.id)
         } yield throw ex
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/backendforfrontend/api/impl/EServicesApiServiceImpl.scala]:
```suggestion
        } yield Future.failed(ex)
```

**Review by galales — APPROVED** [review]
🚀 
one more `throw` to remove, approving in advance

**Review by beetlecrunch — APPROVED** [review]

**Pull request merged**